engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -2208,9 +2208,11 @@ private void advanceStop(final VMInstanceVO vm, final boolean cleanUpEvenIfUnabl
 
             boolean result = stateTransitTo(vm, Event.OperationSucceeded, null);
             if (result) {
+                vm.setPowerState(PowerState.PowerOff);
+                _vmDao.update(vm.getId(), vm);
                 if (VirtualMachine.Type.User.equals(vm.type) && ResourceCountRunningVMsonly.value()) {
                     ServiceOfferingVO offering = _offeringDao.findById(vm.getId(), vm.getServiceOfferingId());
-                    resourceCountDecrement(vm.getAccountId(),new Long(offering.getCpu()), new Long(offering.getRamSize()));
+                    resourceCountDecrement(vm.getAccountId(), offering.getCpu().longValue(), offering.getRamSize().longValue());
```

Comment:
in the database these fields are nullable. Are we sure these are not potential NPE´s?

Reply:
The code below gets cpu/ram from user_vm_details if it is null in service offering.

ServiceOfferingVO offering = _offeringDao.findById(vm.getId(), vm.getServiceOfferingId());

Reply:
reverting this change.

Reply:
As Wei said, it's handled in `OfferingDao`'s `findById`.

https://github.com/apache/cloudstack/blob/a31449b10490cb386d60fc961f6d3bf3bdb09ccd/engine/schema/src/main/java/com/cloud/service/dao/ServiceOfferingDaoImpl.java#L171-L182

https://github.com/apache/cloudstack/blob/a31449b10490cb386d60fc961f6d3bf3bdb09ccd/engine/schema/src/main/java/com/cloud/service/dao/ServiceOfferingDaoImpl.java#L205-L219

Reply:
`new Long(null)` also throws a NPE. So, this is not really relevant to this PR but a minor code improvement.

Reply:
yes, the method has been used in many places, it works well.
No need to add a null check I think.

anyway , good question @DaanHoogland

Reply:
ok, guys, I agree i think, if we get an NPE a check would not help us take measure against it. decreasing by 0 (zero) does not make sense either.

Any improvement on this would be out of scope for this PR and I think it might not be high prio (just good to keep in mind)